Fix async calls in BaseSelectItem

Codinkat opened the pull request:
As the title says, fixed async calls in some base components. Also made some improvements to keep implementations details inside the class, state management, event catch-and-modify, etc.

mikes-gh commented:
This is going to conflict.  https://github.com/Garderoben/MudBlazor/pull/486/ I would suggest wait till this is in then rebase

Codinkat commented:
@mikes-gh Ok I took a look on #486. That's right, I think it should be merged before this one. Once done, I'll review mine accordingly.

Codinkat commented:
Ok, so for simplicity I reverted the changes to MudBasePicker (and renamed this PR), so it could be at least possible to merge the changes for BaseSelectItem.

mikes-gh commented:
You will need to mark this as ready for review if your done.

Codinkat commented:
Haha, yes!

henon commented:
@Codinkat #486 has already been merged. Maybe reset hard the revert and force push? Then I'll merge

Codinkat commented:
@henon When I first submitted the PR, it contained changes for two files. Because BasePicker has conflicting changes, I reverted it so it remains only BaseSelectItem in this PR now and you can merge it. For BasePicker, I will check separately and submit an other PR if needed.
